src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -263,27 +272,23 @@ public Task<DocumentOnAutoInsertResponseItem[]> GetDocumentOnAutoInsertAsync(Doc
             => _requestHandlerProvider.ExecuteRequestAsync<DocumentOnAutoInsertParams, DocumentOnAutoInsertResponseItem[]>(_queue, MSLSPMethods.OnAutoInsertName,
                 autoInsertParams, _clientCapabilities, _clientName, cancellationToken);
 
-        private void DiagnosticService_DiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
+        private void DiagnosticService_DiagnosticsUpdated(object _, DiagnosticsUpdatedArgs e)
+            => DiagnosticService_DiagnosticsUpdated(e.Solution, e.DocumentId);
+
+        private void DiagnosticService_DiagnosticsUpdated(Solution? solution, DocumentId? documentId)
         {
             // LSP doesn't support diagnostics without a document. So if we get project level diagnostics without a document, ignore them.
-            if (e.DocumentId != null && e.Solution != null)
+            if (documentId != null && solution != null)
             {
-                var document = e.Solution.GetDocument(e.DocumentId);
+                var document = solution.GetDocument(documentId);
                 if (document == null || document.FilePath == null)
-                {
                     return;
-                }
 
                 // Only publish document diagnostics for the languages this provider supports.
                 if (document.Project.Language != CodeAnalysis.LanguageNames.CSharp && document.Project.Language != CodeAnalysis.LanguageNames.VisualBasic)
-                {
                     return;
-                }
 
-                // LSP does not currently support publishing diagnostics incrementally, so we re-publish all diagnostics.
-                var asyncToken = _listener.BeginAsyncOperation(nameof(PublishDiagnosticsAsync));
-                Task.Run(() => PublishDiagnosticsAsync(document))
-                    .CompletesAsyncOperation(asyncToken);
+                _diagnosticsWorkQueue.AddWork(document.Id);
```

Comment:
this it the crux of the fix.   Previously we kicked off tasks that would run concurrently (which then mutated shared state). This corrupted that state, leading to be instability and runaway perf issues.  Now, we use the work queue to do this work as the work queue only calls its callback in a serialized fashion.